### PR TITLE
Add support for encryption keys in the protocol

### PR DESCRIPTION
**Pull request by pnarayanan**
Add support for encryption keys in put requests. PutRequest V4 is introduced
for this purpose. This patch does not enable it in the write path. This change
should first go to all the servers so they understand V4. A subsequent patch will
then enable V4 in the write path.

### CODE REVIEW

File: ambry-protocol/src/main/java/com.github.ambry.protocol/PutRequest.java [hunk 1]
```diff
@@ -39,6 +39,7 @@
   protected long sentBytes = 0;
   protected final BlobProperties properties;
   protected final BlobType blobType;
+  protected final byte[] blobKey;
```
[the remark this answers is not in the document]
Comment (vgkholla) [reply ambry-protocol/src/main/java/com.github.ambry.protocol/PutRequest.java]: +1

File: ambry-protocol/src/main/java/com.github.ambry.protocol/PutRequest.java [hunk 2]
```diff
@@ -104,8 +109,16 @@ public long sizeInBytes() {
 
   private int sizeExcludingBlobAndCrcSize() {
     // size of (header + blobId + blob properties + metadata size + metadata + blob size + blob type)
-    return (int) super.sizeInBytes() + blobId.sizeInBytes() + BlobPropertiesSerDe.getBlobPropertiesSerDeSize(properties)
-        + UserMetadata_Size_InBytes + usermetadata.capacity() + Blob_Size_InBytes + BlobType_Size_InBytes;
+    int sizeToReturn =
+        (int) super.sizeInBytes() + blobId.sizeInBytes() + BlobPropertiesSerDe.getBlobPropertiesSerDeSize(properties)
+            + UserMetadata_Size_InBytes + usermetadata.capacity() + Blob_Size_InBytes + BlobType_Size_InBytes;
+    if (getVersionId() == Put_Request_Version_V4) {
```
Comment (vgkholla) [inline ambry-protocol/src/main/java/com.github.ambry.protocol/PutRequest.java]: todo: remove this check also?

File: ambry-protocol/src/main/java/com.github.ambry.protocol/PutRequest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (pnarayanan) [reply ambry-protocol/src/main/java/com.github.ambry.protocol/PutRequest.java]: `byte[]` array is the simplest thing that serves the purpose. I find that `ByteBuffer` is an overkill for this and is unnecessary (maintaining positions and such when not useful is really a burden, especially within tests). I don't think parity is important enough here to overcomplicate the code, but I'd like to know if there is anything that gets simplified by using a `ByteBuffer`

File: ambry-protocol/src/main/java/com.github.ambry.protocol/PutRequest.java [hunk 3]
```diff
@@ -53,10 +54,12 @@
   private static final int UserMetadata_Size_InBytes = 4;
   private static final int Blob_Size_InBytes = 8;
   private static final int BlobType_Size_InBytes = 2;
+  private static final int BlobKeyLength_Size_InBytes = 2;
```
[the remark this answers is not in the document]
Comment (pnarayanan) [reply ambry-protocol/src/main/java/com.github.ambry.protocol/PutRequest.java]: The full caps thing is going to be a sweeping change, and if we do that we should really make it across all of protocol. Do you think we should make that change as part of this patch?

[the remark this answers is not in the document]
Comment (pnarayanan) [reply ambry-protocol/src/main/java/com.github.ambry.protocol/PutRequest.java]: The same goes with `Short.BYTES`, really. But I have made the changes within `PutRequest`. I suppose we should follow the practice of fixing files that we touch.

File: ambry-protocol/src/main/java/com.github.ambry.protocol/PutRequest.java (hunk 2 above)
Comment (pnarayanan) [reply ambry-protocol/src/main/java/com.github.ambry.protocol/PutRequest.java]: fixed.

**Comment by pnarayanan** [comment]
@nsivabalan @vgkholla addressed comments.

**Comment by nsivabalan** [comment]
Looks good.

File: ambry-protocol/src/test/java/com.github.ambry.protocol/RequestResponseTest.java [hunk 4]
```diff
@@ -165,33 +241,28 @@ public void putRequestResponseTest() throws IOException {
         new BlobProperties(blobSize, "serviceID", "memberId", "contentType", false, Utils.Infinite_Time,
             Utils.getRandomShort(TestUtils.RANDOM), Utils.getRandomShort(TestUtils.RANDOM));
     testPutRequest(clusterMap, correlationId, clientId, blobId, blobProperties, userMetadata, BlobType.DataBlob, blob,
```
Comment (vgkholla) [inline ambry-protocol/src/test/java/com.github.ambry.protocol/RequestResponseTest.java]: minor: Might be better off to do invalid version test as one of the components in this function than in `testPutRequest`

File: ambry-protocol/src/main/java/com.github.ambry.protocol/PutRequest.java [hunk 5]
```diff
@@ -123,6 +137,14 @@ public long writeTo(WritableByteChannel channel) throws IOException {
         bufferToSend.putInt(usermetadata.capacity());
         bufferToSend.put(usermetadata);
         bufferToSend.putShort((short) blobType.ordinal());
+        // @todo remove this check once V4 is everywhere.
+        if (getVersionId() == PUT_REQUEST_VERSION_V4) {
+          short keyLength = blobKey == null ? 0 : (short) blobKey.length;
```
Comment (vgkholla) [inline ambry-protocol/src/main/java/com.github.ambry.protocol/PutRequest.java]: functionality question: is this allowed to be none? After EAR is implemented, there will still be a situation where something is not encrypted?

File: ambry-protocol/src/main/java/com.github.ambry.protocol/PutRequest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (vgkholla) [reply ambry-protocol/src/main/java/com.github.ambry.protocol/PutRequest.java]: If the conclusions of our offline discussions still hold, let's make this a `ByteBuffer`

File: ambry-protocol/src/main/java/com.github.ambry.protocol/PutRequest.java (hunk 5 above)
Comment (pnarayanan) [reply ambry-protocol/src/main/java/com.github.ambry.protocol/PutRequest.java]: yes, as per our current design, account owners get to enable/disable encryption at container level. @nsivabalan please chime in.

File: ambry-protocol/src/test/java/com.github.ambry.protocol/RequestResponseTest.java (hunk 4 above)
Comment (pnarayanan) [reply ambry-protocol/src/test/java/com.github.ambry.protocol/RequestResponseTest.java]: Doesn't make a lot of difference I think and I just found this convenient.

**Comment by pnarayanan** [comment]
@nsivabalan @vgkholla Addressed comments.

**Comment by pnarayanan** [comment]
@vgkholla addressed comments, verified that gradle test succeeds.

**Pull request merged**